mplex/src/lib.rs

```diff
@@ -18,730 +18,372 @@
+        Multiplex {
+            inner: self.inner.clone(),
         }
-
-        let id = if let Some((id, _)) = lock.to_open.iter().next() {
-            *id
-        } else {
-            return Ok(Async::NotReady);
-        };
-
-        let name = lock.to_open.remove(&id).expect(
-            "We just checked that this key exists and we have exclusive access to the map, QED",
-        );
-
-        lock.open_stream(id);
-
-        Ok(Async::Ready(Some(Substream::new(
-            id,
-            self.end,
-            name,
-            Arc::clone(&self.state),
-        ))))
     }
 }
 
-pub struct OutboundFuture<T, Buf: Array> {
-    meta: Arc<MultiplexMetadata>,
-    current_id: Option<(io::Cursor<ByteBuf>, u32)>,
-    state: Arc<Mutex<MultiplexShared<T, Buf>>>,
+// Struct shared throughout the implementation.
+struct MultiplexInner<C> {
+    // Underlying stream.
+    inner: Fuse<Framed<C, codec::Codec>>,
+    // Buffer of elements pulled from the stream but not processed yet.
+    buffer: Vec<codec::Elem>,
+    // List of Ids of opened substreams. Used to filter out messages that don't belong to any
+    // substream.
+    opened_substreams: FnvHashSet<u32>,
+    // Id of the next outgoing substream. Should always increase by two.
+    next_outbound_stream_id: u32,
+    // List of tasks to notify when a new element is inserted in `buffer`.
+    to_notify: Vec<task::Task>,
 }
 
-impl<T, Buf: Array> OutboundFuture<T, Buf> {
-    fn new(muxer: BufferedMultiplex<T, Buf>) -> Self {
-        OutboundFuture {
-            current_id: None,
-            meta: muxer.meta,
-            state: muxer.state,
-        }
+// Processes elements in `inner` until one matching `filter` is found.
+//
+// If `NotReady` is returned, the current task is scheduled for later, just like with any `Poll`.
+// `Ready(Some())` is almost always returned. `Ready(None)` is returned if the stream is EOF.
+fn next_match<C, F, O>(inner: &mut MultiplexInner<C>, mut filter: F) -> Poll<Option<O>, IoError>
+where C: AsyncRead + AsyncWrite,
+      F: FnMut(&codec::Elem) -> Option<O>,
+{
+    if let Some((offset, out)) = inner.buffer.iter().enumerate().filter_map(|(n, v)| filter(v).map(|v| (n, v))).next() {
+        inner.buffer.remove(offset);
+        return Ok(Async::Ready(Some(out)));
     }
-}
-
-fn nonce_to_id(id: usize, end: Endpoint) -> u32 {
-    id as u32 * 2 + if end == Endpoint::Dialer { 0 } else { 1 }
-}
 
-impl<T: AsyncWrite, Buf: Array> Future for OutboundFuture<T, Buf> {
-    type Item = Option<Substream<T, Buf>>;
-    type Error = io::Error;
-
-    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let mut lock = match self.state.poll_lock() {
-            Async::Ready(lock) => lock,
-            Async::NotReady => return Ok(Async::NotReady),
+    loop {
+        let elem = match inner.inner.poll() {
+            Ok(Async::Ready(item)) => item,
+            Ok(Async::NotReady) => {
+                inner.to_notify.push(task::current());
```

Comment:
Why is the manual task management necessary here? Is the `inner.inner` stream not responsible for setting things up such that the task is notified when progress can be made? After all `futures::stream::Stream`'s documentation states that "If `NotReady` is returned then this stream's next value is not ready yet and implementations will ensure that the current task will be notified when the next value may be ready."

Reply:
If we call `next_match` from task A then later from task B, only task B will be awakened.
Usually futures and streams are owned by one specific task, so it's not a problem, but here it's not the case.

Reply:
Ah yes, thanks! I think this is documented in `Future::poll`:

> Note that if NotReady is returned it only means that this task will receive a notification. Historical calls to poll with different tasks will not receive notifications. In other words, implementers of the Future trait need not store a queue of tasks to notify, but only the last task that called this method. Alternatively callers of this method can only rely on the most recent task which call poll being notified when a future is ready.